File: C++/src/3sum.cc
#include "3sum.hpp"
#include <algorithm>

vector<vector<int>> ThreeSum::Solution::twoSum(int left, vector<int> &nums,
                                               int target) {
  int right = nums.size() - 1;
  vector<vector<int>> res;

  while (left < right) {
    int sum = nums[left] + nums[right];

    if (sum == target) {
      res.push_back({nums[left], nums[right]});
      left++;
      while (left < right && nums[left - 1] == nums[left]) {
        left++;
      }
    } else if (sum < target) {
      left++;
    } else {
      right--;
    }
  }

  return res;
}

vector<vector<int>> ThreeSum::Solution::threeSum(vector<int> &nums) {
  sort(nums.begin(), nums.end());

  vector<vector<int>> res;

  for (size_t i = 0; i < nums.size(); i++) {
    if (nums[i] > 0) {
      break;
    }
    if (i > 0 && nums[i] == nums[i - 1]) {
      continue;
    }
    auto pairs = twoSum(i + 1, nums, -nums[i]);

    for (auto pair : pairs) {
      res.push_back({nums[i], pair[0], pair[1]});
    }
  }

  return res;
}

File: JS/src/3sum.js
/**
 * @param {number} left
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
function twoSum(left, nums, target) {
  const result = [];
  let right = nums.length - 1;

  while (left < right) {
    const sum = nums[left] + nums[right];

    if (sum === target) {
      result.push([nums[left], nums[right]]);
      left++;
      while (left < right && nums[left] === nums[left - 1]) {
        left++;
      }
    } else if (sum < target) {
      left++;
    } else {
      right--;
    }
  }

  return result;
}

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
function threeSum(nums) {
  nums = nums.sort((a, b) => a - b);

  const res = [];

  for (let i = 0; i < nums.length; i++) {
    if (nums[i] > 0) {
      break;
    }
    if (i > 0 && nums[i] === nums[i - 1]) {
      continue;
    }

    const pairs = twoSum(i + 1, nums, -nums[i]);
    for (const pair of pairs) {
      res.push([nums[i], ...pair]);
    }
  }

  return res;
}

export { threeSum };
